src/helpers/blob.rs

```diff
@@ -0,0 +1,240 @@
+/// regular byte slice (`[u8]`) since it's mostly constructed from `Box<[u8]>`
+/// via C FFI.
+///
+/// It's worth noting that heap allocation will often align to pointer size, so
+/// no unaligned load should happen once the value is constructed from
+/// heap-allocated bytes.
+#[repr(C, packed)]
+pub struct Blob<T: BlobLayout>(T::Header, [u8]);
+
+/// Marker trait for dynamic struct layouts prefixed with `Self::Header` type
+/// of a statically-known size. Used in tandem with `Blob`.
+pub trait BlobLayout {
+    type Header: AsBytes + FromBytes;
+}
+
+impl<'a, T: BlobLayout> Blob<T> {
+    pub fn header(&self) -> &T::Header {
+        // SAFETY: The only way to construct `Blob` is via
+        // `Self::from_boxed`, which requires that the source reference is
+        // aligned at least as `T::Header` and since `Blob` is
+        // `#[repr(C)]` (and so is `T::Header`, because it  implements `AsBytes`
+        // which requires which requires being `#[repr(C)]`), so the reference
+        // to its first field will be aligned at least as `T::Header`.
+        unsafe { &self.0 }
+    }
+
+    pub(crate) fn tail(&self) -> &[u8] {
+        &self.1
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        AsBytes::as_bytes(self)
+    }
+
+    pub fn into_bytes(self: Box<Self>) -> Box<[u8]> {
+        AsBytes::into_bytes(self)
+    }
+
+    pub fn from_boxed(boxed: Box<[u8]>) -> Box<Self> {
+        FromBytes::from_boxed(boxed)
+    }
+
+    pub(crate) unsafe fn ref_cast<U: BlobLayout>(&self) -> &Blob<U> {
+        Blob::<U>::from_bytes(self.as_bytes())
+    }
+}
+
+// SAFETY: The entire struct is `#[repr(C)]` and so is the header (because it
+// implements `AsBytes` as well)
+unsafe impl<T: BlobLayout> AsBytes for Blob<T> {}
+
+unsafe impl<T: BlobLayout> FromBytes for Blob<T> {
+    // Require that the allocation is at least as aligned as its header to
+    // safely reference it as the first field. (despite
+    // `Blob` being technically `#[repr(packed)]`)
+    unsafe fn min_layout() -> std::alloc::Layout {
+        std::alloc::Layout::new::<T::Header>()
+    }
+
+    unsafe fn ptr_cast(source: *const [u8]) -> *const Self {
+        assert_ne!(source as *const (), std::ptr::null());
+        // SAFETY: [u8] is 1-byte aligned so no need to check before deref
+        let len = (&*source).len();
+        let tail_len = len - std::mem::size_of::<T::Header>();
+
+        // SAFETY: This assumes that the pointer to slices and slice-based DSTs
+        // have the same metadata (verified by the compiler at compile-time)
+        let slice = std::slice::from_raw_parts(source as *const T::Header, tail_len);
+        slice as *const [T::Header] as *const _
+    }
+}
+
+/// Defines a trait for accessing dynamic fields (byte slices) for structs that
+/// have a header of a known size which also defines the rest of the struct
+/// layout.
+/// Assumes a contiguous byte buffer.
+#[macro_export]
+macro_rules! blob {

```

Comment:
This is very clever. The macro definition is not so easy to read, but its usage is really nice. I like it!